Websites (#10)
mine-cetinkaya-rundel/quarto-tdg: websites.qmd

![Page in the Reference](images/website-hybrid-reference.png){fig-alt="A screenshot of a page in the Reference. A box highlights the Reference item in the top navigation bar and the sidebar navigation that is titled Reference."}
:::

Hybrid navigation on the Quarto documentation site. Guide pages share sidebar navigation, which is different to the sidebar on Reference pages.

:::

To implement hybrid navigation you combine a `navbar` with an array of `sidebar`s, linking the `text` of `navbar` items to the `title` of the `sidebar` items.

For example, you might have a `navbar` that links to Teaching and Research sections:

```{.yaml filename="_quarto.yml"}
website:
  title: Drake Lab
  navbar:
    left:
      - text: "Research"            
        href: research.qmd          
      - text: "Teaching"
        href: teaching.qmd
```

For each section to have its own sidebar, `sidebar` should be an array (i.e. elements prefaced with `-`), and the `title` of each item should match the `text` from an item in the `navbar`:

> ```{.yaml filename="_quarto.yml"}

Author: Yes, and I vote we omit unless neccessary.

> ```{.yaml filename="_quarto.yml"}

Reviewer: Works for me!

You then check `_freeze/`, but not `_site/`, into version control.
Your CI agent then uses the Quarto CLI to render and publish your site.

The most common CI service for this approach is GitHub. 
The project source is stored in a GitHub repo, and a [GitHub Action](#github-actions) is used to setup Quarto, 
and run the required Quarto commands.
The Quarto command could be as simple as `quarto publish {venue}`, which allows your publishing venue to be any of those supported by `quarto publish`.

### Remote code execution and remote render

Remote code execution combined with remote rendering provides the most protection for keeping your work reproducible---your CI will fail as soon as it isn't.
However, it comes with the most setup, maintenance, and cost (time and potentially money) since all the work is happening on a remote server.

You now only check your source documents (e.g. `.qmd` files) into version control. 
Your CI service needs to install your computational environment, so you likely will also create and check-in a record of the packages required (e.g. using `renv` for R, or `requirements.txt`, `pipenv`, or `uv` for Python).  

> When your CI runs, it will install your computational environment, install Quarto, and then run a full render and publish.  

Reviewer: Wonder why this sentence is on the next line in the rendered book?

Author: Fixed.

        with:                                       #<8>
          target: gh-pages                          #<8>
        env:                                        #<8> 
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} #<8>
```

1. Describes when this workflow should trigger. In this case, on any pushes to the `main` branch, or manually (`workflow_dispatch`).
2. A name for the workflow displayed in GitHub Action's UI.
3. A workflow can have one or more `jobs` that run in parallel. Most Quarto CI workflows have a single job.
4. A job identifier and settings including permissions and the type of machine the job runs on. This workflow needs `write` permissions because it will write to the `gh-pages` branch to publish the site.
5. The meat of the job is a series of steps to be run in order.
6. Check out the project repository onto the machine.
7. Install the Quarto CLI. Uses the latest release by default.
8. Runs `quarto publish {target}` (by default includes a full `quarto render`). Depending on your `target` you'll need to set some environment variables with your credentials on your publishing venue.

A GitHub Action that publishes to GitHub Pages

:::

> Here's three common ways you might customize this workflow:

Reviewer: ```suggestion
Three common ways you might customize this workflow are as follows:
```

* Split rendering and publishing. If you are publishing to a venue that isn't supported by `quarto publish`, you might split up the "Render and Publish" step. You could use the `quarto-dev/quarto-actions/render` action to render your site. Then add a custom step for the deployment, using actions from someone like Netlify, or your own custom commands. 

:::::{.column-page layout-ncol="2"}

::: {#lst-language-setup-r}
```{.yaml filename=".github/workflows/publish.yml"}
      - name: Install R
        uses: r-lib/actions/setup-r@v2
        with:
          r-version: '4.4.3'

      - name: Install R Dependencies
        uses: r-lib/actions/setup-renv@v2
        with:
          cache-version: 1
```

Setup R
:::

::: {#lst-language-setup-python}

```{.yaml filename=".github/workflows/publish.yml"}
      - name: Install Python and Dependencies
        uses: actions/setup-python@v5
        with:
          python-version: '3.13'
          cache: 'pip'

      - run: pip install jupyter
      - run: pip install -r requirements.txt
```
Setup Python
:::

:::::

## Wrapping Up

> In this chapter, you've learnt about the key components of a Quarto website: content, navigation and appearance. 

Reviewer: ```suggestion
In this chapter, you learned about the key components of a Quarto website: content, navigation, and appearance. 
```

> You've also learnt about a few techniques for publishing your site.

Reviewer: ```suggestion
You also learned about a few techniques for publishing your site.
```

> But, there's still a lot of features we haven't talked about.

Reviewer: ```suggestion
But, there are many other features we haven't talked about.
```

Some notable ones include:

> * Drafts: set `draft: true` to be able to preview pages, but not have them included in a site render. <https://quarto.org/docs/websites/website-drafts.html>

Reviewer: ```suggestion
* Drafts: Set `draft: true` to be able to preview pages, but not have them included in a site render, <https://quarto.org/docs/websites/website-drafts.html>.
```

> * Comments: integrate third-party tools to allow on page comments. <https://quarto.org/docs/output-formats/html-basics.html#commenting> 

Reviewer: ```suggestion
* Comments: Integrate third-party tools to allow your readers to leave comments on your pages, <https://quarto.org/docs/output-formats/html-basics.html#commenting>.
```

> * Announcements: add a bar to the top of the site for crucial info. <https://quarto.org/docs/websites/website-tools.html#announcement-bar>

Reviewer: ```suggestion
* Announcements: Add a bar to the top of the site for crucial info, <https://quarto.org/docs/websites/website-tools.html#announcement-bar>.
```

You can find comprehensive lists of all your options are in the Reference in the Quarto documentation:

> * You find website-level options in [Projects > Websites](https://quarto.org/docs/reference/projects/websites.html)

Reviewer: ```suggestion
* You find website-level options in [Projects > Websites](https://quarto.org/docs/reference/projects/websites.html).
```

> * You find page-level options (that could be set site-wide in `_quarto.yml`) in [Formats > HTML](https://quarto.org/docs/reference/formats/html.html)

Reviewer: ```suggestion
* You find page-level options (that could be set site-wide in `_quarto.yml`) in [Formats > HTML](https://quarto.org/docs/reference/formats/html.html).
```
